### ISSUE TITLE
Repairs get stuck in sidecar mode

### ISSUE THREAD

**Issue by kzemek** [thelastpickle/cassandra-reaper#961]
I'm running Reaper 2.0.5 with multi-datacenter Cassandra, with a Reaper instance colocated with each Cassandra node and `REAPER_DATACENTER_AVAILABILITY=SIDECAR` . The repairs do complete as expected, but they seem to take a (too) long time, with `last event` in the GUI often showing `Postponed repair segment <UUID> on repair run with id <UUID> because we couldn't get hosts metrics on <IP>`.  In logs of Reaper instances I get a lot of lines like `INFO   [2020-09-17 13:01:39,683] [pool-13-thread-7] i.c.s.SegmentRunner - Trying to get metrics from remote DCs for <IP> in <DC> of <CLUSTER>`.

Does Reaper slow down its repairs because it tries to repair other nodes even though it's in sidecar mode, or am I just misinterpreting log messages?

**Comment by kzemek** [comment]
Alright, I think I was just confused because my repairs were stuck, and all I could find were these messages. As far as I understand it the messages are not about Reaper instance contacting other Cassandras, but rather contacting other Reaper instances via the backend store. My repairs were fixed by restarting all of the Reaper instances (just restarting repairs didn't help, the new repairs were stuck as well).

I'm still getting a lot of `we couldn't get hosts metrics on` messages though, even though all the instances seem to be able to share data and eventually the errors are resolved - is there a timeout for that I can tweak?

**Comment by kzemek** [comment]
I managed to finish a repair of my keyspaces - it got stuck multiple times in the process, for hours at a time with no segments running. Restarting all Reaper containers has helped each time. For an example stuck repair, the logs (filtered out `Responded to metric` and `Got metric request` messages) only had the following messages between the last repaired segment and Reaper shutdown:

```
[2020-09-17 20:29:46,511] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe:9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.SegmentRunner - Repair command 4240 on segment 9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe returned with state DONE
[2020-09-17 20:29:46,890] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe:9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.CassandraStorage - Trying to release lead on segment 9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe for instance 9b7701ae-7c55-4152-9dac-b5b37b3be84b
[2020-09-17 20:29:46,891] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe:9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.CassandraStorage - Released lead on segment 9a27cad7-f8f0-11ea-bfa4-e54143dc5bbe
[2020-09-17 20:30:11,491] [cluster-name:46080c70-f8f5-11ea-a728-f774b92e9f88] i.c.s.RepairRunner - Running segment for range (-9217940801891802122,-9217940801891802122]
[2020-09-17 20:30:11,528] [cluster-name:46080c70-f8f5-11ea-a728-f774b92e9f88] i.c.s.RepairRunner - Next segment to run : 460d63c2-f8f5-11ea-a728-f774b92e9f88
[2020-09-17 20:30:11,939] [cluster-name:46080c70-f8f5-11ea-a728-f774b92e9f88:460d63c2-f8f5-11ea-a728-f774b92e9f88] i.c.s.SegmentRunner - Ok to repair segment '460d63c2-f8f5-11ea-a728-f774b92e9f88' on repair run with id '46080c70-f8f5-11ea-a728-f774b92e9f88'
[2020-09-17 20:30:12,482] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.RepairRunner - Running segment for range (-9217940801891802122,-9217940801891802122]
[2020-09-17 20:30:12,516] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.RepairRunner - Next segment to run : 9a286701-f8f0-11ea-bfa4-e54143dc5bbe
[2020-09-17 20:30:12,875] [cluster-name:9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe:9a286701-f8f0-11ea-bfa4-e54143dc5bbe] i.c.s.SegmentRunner - Ok to repair segment '9a286701-f8f0-11ea-bfa4-e54143dc5bbe' on repair run with id '9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe'
[2020-09-17 20:40:53,762] [ReaperApplication-scheduler] i.c.ReaperApplication - Purged 0 repair runs from history
[2020-09-17 21:40:53,914] [ReaperApplication-scheduler] i.c.ReaperApplication - Purged 0 repair runs from history
[2020-09-17 22:40:54,052] [ReaperApplication-scheduler] i.c.ReaperApplication - Purged 0 repair runs from history
[2020-09-17 22:49:46,534] [Thread-14] o.e.j.s.AbstractConnector - Stopped application@6d293993{HTTP/1.1,[http/1.1]}{127.0.0.1:8080}
```

Another thing I noticed that might be relevant is that from Reaper's point of view each processed segment is running on `127.0.0.1`, so even though I have 4 threads set for repair, it repaired at most 1 segment at a time (`because segment <UUID> is running on host 127.0.0.1`).

Note that in this case I'm running two simultaneous repairs (`46080c70-f8f5-11ea-a728-f774b92e9f88` and `9a1f1830-f8f0-11ea-bfa4-e54143dc5bbe`), but I only started the second one after the first has already got stuck at least once.

After the initial stages of the repair, `we couldn't get hosts metrics on` did not show up again, so it might not be relevant after all.

**Comment by adejanovski** [comment]
Hi @kzemek,

I'm not happy with how grabbing the metrics performs in distributed modes either. We're currently reworking it to make repair much more efficient.

The number of threads is only relevant for repairing Cassandra 2.2. All more recent versions of Cassandra will turn multi token ranges repairs in a single repair session with Reaper, thus requiring a single thread on the Cassandra side.

The number of potential concurrent repairs when using reaper will be `[ number_of_nodes / RF ]` (with RF being the total number of replicas in the cluster, which means that if you have two DCs with RF 3 both, it'll be 6).
We've seen Reaper 2.0+ struggling to match the number of concurrent requests because of implementation flaws.
It is being reworked as well and I'll update this issue once we have merged the changes.

**Comment by adejanovski** [comment]
The newly released 2.2 should fix this issue. Please reopen if it's not the case.
